plugin/src/main/java/com/playmonumenta/scriptedquests/zones/zone/Zone.java

```diff
@@ -174,25 +181,159 @@ public boolean splitByOverlap(BaseZone overlap, Zone<T> otherZone) {
 	 *
 	 * This works with only one zone's fragments at a time, and doesn't
 	 * need to be run again. This reduces n significantly for runtime.
-	 *
-	 * TODO Implement this based on the same function in:
-	 * https://github.com/NickNackGus/monumenta-zone-prototype/blob/master/python/lib/zone/zone.py
-	 * Also needs to invalidate fragments of merged zones.
 	 */
-	/*
 	public void defragment() {
 		if (mFragments.size() < 2) {
 			return;
 		}
 
-		HashMap<int, HashMap<LinkedHashSet<int>, ZoneFragment>> allMergedCombinations = new HashMap<int, HashMap<LinkedHashSet<int>, ZoneFragment>>;
-		private ArrayList<ZoneFragment> recursiveOptimalDefrag(HashMap<int, HashMap<LinkedHashSet<int>, ZoneFragment>> allMergedCombinations,
-		                                                       ArrayList<ZoneFragment> resultsSoFar,
-		                                                       LinkedHashSet<int> remainingIds);
+		class Defragmenter<T> {
```

Comment:
This is pretty big. Can we move this to a separate file? (ZoneDefragmenter.java ?)